### PR TITLE
feat: unsaved change popup

### LINKED ISSUE #2207: [feature]: Unsaved changes warning popup
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Auto-save or show popup when detect any change of api (paramter, body,...)

### Why should this be worked on?

I am developing my API is evolving. When I edit the parameter, body of the api and mistakenly click to another api, the edited content is lost.
I think it's better if have function auto-save or show popup when detect any change of api (paramter, body,...)
It will save time when working with multiple APIs

### PR DESCRIPTION
**Pull request by nivedin**
Closes #2207 

### Description
Shows a popup when there are any changes done in the current request (API, parameter, body,..) after clicking on any other collection request.

### Checks
- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### Additional Information
When a user is working on a request and made many changes and accidentally clicks on another collection request all the changes that have been done will be lost. So to prevent this problem this PR adds a function that triggers a popup when any changes are detected when the user clicks on another collection request which asks whether to save or discard the changes.

**TODO**
- [x] Team > Request
- [x] History

### CODE REVIEW

**Review by liyasthomas — APPROVED** [review]

File: packages/hoppscotch-app/components/collections/my/Request.vue [hunk 1]
```diff
@@ -143,129 +138,290 @@
...
+    } else {
+      setRESTSaveContext(null)
+    }
+  }
+}
+
+// Save current request to the collection
+const saveRequestChange = () => {
+  const saveCtx = getRESTSaveContext()
+  saveCurrentRequest(saveCtx)
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/my/Request.vue:354]: Type Error over here.

`saveCtx` can be null.

File: packages/hoppscotch-app/components/collections/my/Request.vue [hunk 2]
```diff
@@ -143,129 +138,290 @@
...
+      setRestReq(props.request)
+      toast.success(`${t("request.saved")}`)
+    } catch (e) {
+      setRESTSaveContext(null)
+      saveCurrentRequest(saveCtx)
+    }
+  } else if (saveCtx.originLocation === "team-collection") {
+    const req = getRESTRequest()
+
+    // TODO: handle error case (NOTE: overwriteRequestTeams is async)
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/my/Request.vue:404]: This TODO is implemented and can be removed.

File: packages/hoppscotch-app/components/collections/my/Request.vue [hunk 3]
```diff
@@ -143,129 +138,290 @@
...
+const dragging = ref<boolean>(false)
+const requestMethodLabels = {
+  get: "text-green-500",
+  post: "text-yellow-500",
+  put: "text-blue-500",
+  delete: "text-red-500",
+  default: "text-gray-500",
+}
+const confirmRemove = ref<boolean>(false)
+const confirmApiChange = ref<boolean>(false)
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/my/Request.vue:244]: `confirmChange` is a better name as we are reserving API naming for things in the future.

File: packages/hoppscotch-app/components/collections/teams/Request.vue [hunk 4]
```diff
@@ -143,116 +139,266 @@
...
+
+const active = useReadonlyStream(restSaveContext$, null)
+
+const isSelected = computed<boolean>(() => {
+  return (
+    props.picked &&
+    props.picked.pickedType === "team-collection" &&
+    props.picked.requestID === props.requestIndex
+  )
+})
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/teams/Request.vue:251]:
```suggestion
const isSelected = computed(() =>
    !!props.picked &&
    props.picked.pickedType === "team-collection" &&
    props.picked.requestID === props.requestIndex
)
```

File: packages/hoppscotch-app/components/collections/teams/Request.vue [hunk 5]
```diff
@@ -143,116 +139,266 @@
...
+
+const isActive = computed<boolean>(() => {
+  return (
+    active.value &&
+    active.value.originLocation === "team-collection" &&
+    active.value.requestID === props.requestIndex &&
+    isEqual(active.value.req, props.request)
+  )
+})
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/teams/Request.vue:260]:
```suggestion
const isActive = computed(() =>
    !!active.value &&
    active.value.originLocation === "team-collection" &&
    active.value.requestID === props.requestIndex &&
    isEqual(active.value.req, props.request)
)
```

File: packages/hoppscotch-app/components/collections/teams/Request.vue [hunk 6]
```diff
@@ -143,116 +139,266 @@
...
+
+const dragStart = ({ dataTransfer }) => {
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/teams/Request.vue:262]: Type Error!

File: packages/hoppscotch-app/components/collections/teams/Request.vue [hunk 7]
```diff
@@ -143,116 +139,266 @@
...
+  emit("remove-request", {
+    collectionIndex: props.collectionIndex,
+    folderName: props.folderName,
+    requestIndex: props.requestIndex,
+  })
+}
+
+const getRequestLabelColor = (method: string) => {
+  return (
+    requestMethodLabels[method.toLowerCase()] || requestMethodLabels.default
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/teams/Request.vue:277]: Type Error!

File: packages/hoppscotch-app/components/collections/teams/Request.vue [hunk 8]
```diff
@@ -143,116 +139,266 @@
...
+
+const setRestReq = (request: any) => {
+  setRESTRequest(
+    safelyExtractRESTRequest(
+      translateToNewRequest(request),
+      getDefaultRESTRequest()
+    ),
+    {
+      originLocation: "team-collection",
+      requestID: props.requestIndex,
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/collections/teams/Request.vue:289]: Type Error!

File: packages/hoppscotch-app/locales/en.json [hunk 9]
```diff
@@ -121,6 +122,7 @@
     "team_collections": "Team Collections"
   },
   "confirm": {
+    "confirm_api_change": "Are you sure you want to change the api, unsaved changes will be lost.",
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/locales/en.json:125]: We are not editing an API. We are editing a request. There is a difference!

**Review by AndrewBastin — CHANGES_REQUESTED** [review]
Please make sure your type checking is working correctly, there are so many type errors in your implementation. Make sure you have Vetur (VS Code extension) disabled/uninstalled and Volar (https://marketplace.visualstudio.com/items?itemName=johnsoncodehk.volar) installed and setup properly.

Also, rename everything named with "API", we are not working with an API here, we are working with a request, there is a difference in wording. So you will have to update couple of variables you have declared, the i18n strings added and rename `ApiChangeConfirmModal` to either `ReqChangeConfirmModal` or just `ChangeConfirmModal`.

**Review by liyasthomas — APPROVED** [review]

**Pull request merged**